tests/syntax/test_specifiers.py

```diff
@@ -47,181 +52,501 @@ def test_lazy_value_in_requirement():
     ego = sampleEgoFrom(
-        'ego = Object at 10 @ 5\n'
-        'ego = Object beyond 4 @ 13 by 10 @ 5'
+        'ego = new Object at 10 @ 5\n'
+        'ego = new Object beyond 4 @ 13 by 10 @ 5'
     )
-    assert tuple(ego.position) == pytest.approx((9, 23))
+    assert tuple(ego.position) == pytest.approx((9, 23, 0))
+
+def test_beyond_3d():
+    ego = sampleEgoFrom("""
+        import math
+        ego = new Object at (10, 5, 15)
+        ego = new Object beyond (11, 6, 15 + math.sqrt(2)) by (0, 10, 0)
+    """)
+    assert tuple(ego.position) == pytest.approx((16, 11, 15 + math.sqrt(2) + (10 / math.sqrt(2))))
 
 def test_beyond_no_ego():
-    with pytest.raises(RuntimeParseError):
-        compileScenic('ego = Object beyond 10 @ 10 by 5')
+    with pytest.raises(InvalidScenarioError):
+        compileScenic('ego = new Object beyond 10 @ 10 by 5')
 
 def test_beyond_from():
-    ego = sampleEgoFrom('ego = Object beyond 5 @ 0 by 20 from 5 @ 10')
-    assert tuple(ego.position) == pytest.approx((5, -20))
-    ego = sampleEgoFrom('ego = Object beyond 5 @ 0 by 15 @ 20 from 5 @ 10')
-    assert tuple(ego.position) == pytest.approx((-10, -20))
+    ego = sampleEgoFrom('ego = new Object beyond 5 @ 0 by 20 from 5 @ 10')
+    assert tuple(ego.position) == pytest.approx((5, -20, 0))
+    ego = sampleEgoFrom('ego = new Object beyond 5 @ 0 by 15 @ 20 from 5 @ 10')
+    assert tuple(ego.position) == pytest.approx((-10, -20, 0))
 
+def test_beyond_from_3d():
+    ego = sampleEgoFrom("""
+        import math
+        ego = new Object beyond (11, 6, 15 + math.sqrt(2)) by (0, 10, 0) from (10, 5, 15)
+    """)
+    assert tuple(ego.position) == pytest.approx((16, 11, 15 + math.sqrt(2) + (10 / math.sqrt(2))))
+
+# Visible
 def test_visible():
     scenario = compileScenic(
-        'ego = Object at 100 @ 200, facing -45 deg,\n'
+        'ego = new Object at 100 @ 200, facing -45 deg,\n'
         '             with visibleDistance 10, with viewAngle 90 deg\n'
-        'ego = Object visible'
+        'ego = new Object visible'
     )
     for i in range(30):
         scene = sampleScene(scenario, maxIterations=50)
         ego, base = scene.objects
         assert ego.position.distanceTo(base.position) <= 10
         assert ego.position.x >= base.position.x
         assert ego.position.y >= base.position.y
-
+    
 def test_visible_no_ego():
-    with pytest.raises(RuntimeParseError):
-        compileScenic('ego = Object visible')
+    with pytest.raises(InvalidScenarioError):
+        compileScenic('ego = new Object visible')
 
 def test_visible_from_point():
     scenario = compileScenic(
-        'x = Point at 300@200, with visibleDistance 2\n'
-        'ego = Object visible from x'
+        'x = new Point at 300@200, with visibleDistance 2\n'
+        'ego = new Object visible from x'
     )
-    for i in range(30):
-        scene = sampleScene(scenario, maxIterations=1)
+    for i in range(20):
+        scene = sampleScene(scenario, maxIterations=10)
```

Comment:
Why does this take more than one iteration?

Reply:
Before the visibility heuristics with a very low ray density (which we hadn't settled on at the time). I thought there might be a chance this object failed the visibility check. With the new heuristics this should always pass.

Reply:
Rejection sampling a MeshVolumeRegion can rarely fail, so we allow for multiple iterations in that case.